File: agentkit/sdk/agentkit/infra/events/protocols.py
"""Structural protocols for AgentKit event infrastructure.

Using ``typing.Protocol`` lets ``AgentRuntime`` and other consumers accept both
the in-memory and Redis-backed implementations without an explicit inheritance
relationship, keeping the infra-free core independent of Redis.
"""

from __future__ import annotations

from typing import AsyncGenerator, Protocol, runtime_checkable

from agentkit.infra.events.schemas import AgentKitEvent


@runtime_checkable
class EventBusProtocol(Protocol):
    """Structural interface shared by :class:`~agentkit.infra.events.bus.InMemoryEventBus`
    and :class:`~agentkit.infra.events.bus.EventBus` (Redis-backed).

    Any object that implements ``publish`` and ``subscribe`` with these
    signatures satisfies this protocol.
    """

    async def publish(self, event: AgentKitEvent) -> str:
        """Emit an event; return an opaque message/position ID."""
        ...

    async def subscribe(self, session_id: str) -> AsyncGenerator[AgentKitEvent, None]:
        """Yield events for *session_id*, starting from backlog then live."""
        ...
        # This unreachable yield makes the return type an AsyncGenerator,
        # satisfying Protocol static-analysis requirements.
        yield  # type: ignore[misc]

File: agentkit/ts/src/events/protocols.ts
/**
 * Structural contract for an event bus. The shipped in-memory `EventBus`
 * implements this; users may pass a Redis-, Kafka-, or test-backed bus that
 * satisfies the same shape without inheriting from the concrete class.
 *
 * Mirrors `agentkit.infra.events.protocols.EventBusProtocol` from the Python
 * SDK. The TS variant adds `close()` because the in-memory implementation
 * owns subscriber state that must be released on shutdown.
 *
 * Async/sync flexibility is deliberate: `publish` may return either `void`
 * or an opaque message id (Redis XADD returns one), and `close` may be
 * either sync (in-memory) or async (durable bus needing connection
 * teardown).
 */
import type { AgentKitEvent } from "./schemas";

export interface EventBusProtocol {
  /**
   * Fan an event out to subscribers of its session. May return an opaque
   * message/position id (mirrors Python's `Redis XADD` id) or nothing.
   */
  publish(event: AgentKitEvent): Promise<void | string>;
  /**
   * Subscribe to a session's events. Iterate with `for await`. Implementers
   * MUST implement `.return()` on the returned iterator so consumers can
   * signal early exit and release subscriber resources; the shipped
   * `EventBus.Subscription` does this for in-memory cleanup.
   */
  subscribe(sessionId: string): AsyncIterableIterator<AgentKitEvent>;
  /**
   * Release subscriber state. Idempotent. May be sync (in-memory) or async
   * (durable bus flushing in-flight writes / closing a connection).
   */
  close(): void | Promise<void>;
}
